Extend AssignmentNode to include matrix assignment; fix ArrayNode toTex; add/clean-up unit tests

jakemckown opened the pull request:
Evaluate expressions like:

‘[x, y, z] = [1, 2, 3]’
‘[x; y; z] = [1; 2; 3]’
‘[X[1, 1], X[2, 1]] = [sin(pi), cos(2 * pi)]’

Fixed ArrayNode.toTex() to handle rows and columns correctly, and without the trailing ‘\\\\’ for the last row. Updated the unit tests to reflect the cleaner output. To check why this was necessary, run the current version and this updated version for the following:

‘[x, [y, [z]]]’

jakemckown commented:
Evaluate expressions like:

‘[x, y, z] = [1, 2, 3]’
‘[x; y; z] = [1; 2; 3]’
‘[X[1, 1], X[2, 1]] = [sin(pi), cos(2 * pi)’

Fixed ArrayNode.toTex() to handle rows and columns correctly, and without the trailing ‘\\\\’ for the last row. Updated the unit tests to reflect the cleaner output. To check why this was necessary, run the current version and this updated version for the following:

‘[x, [y, [z]]]’

josdejong commented:
Thanks for your PR Jake. Such a feature can be interesting. I have to give it some more thought to see what the potential is and if this is a common use case that we want to support. I hope to review your PR coming weekend (it's quite large please give me some time ;) ).

Can you in the mean time have a look why some unit tests are failing?

jakemckown commented:
Whoops, I made a change to BlockNode to test the result and never put it back. All unit tests are passing now.

I'm sorry, but now two codecov tests are failing, and I'm unfamiliar with codecov. I'm not sure what it's saying is failing?

jakemckown commented:
Matrix assignment would provide similar functionality to [destructuring in JavaScript](https://javascript.info/destructuring-assignment), a feature also supported by languages like Python and Perl.

Applications like Mathcad also provide matrix assignment, which can be very powerful. Here's a handful of operations in Mathcad that I used to write the unit test I added to parse (the ArrayNode.toTex() function should now properly typeset these as well):

![Matrix Assignment 1](https://user-images.githubusercontent.com/19611724/59376923-f833ad80-8d05-11e9-89d7-e901cf3b0f5e.png)

While operations like this can be very convenient for assigning variables in a compact form, the most powerful feature is destructuring the return value of a matrix-creating function:

![Matrix Assignment 2](https://user-images.githubusercontent.com/19611724/59377240-a63f5780-8d06-11e9-9703-981dc9a15cb5.png)

Users could break large data sets (e.g., pixel values from an image) into smaller pieces and manipulate them separately in different ways. And since matrices can be nested, this can go as many levels deep as needed :)

jakemckown commented:
Regarding the changes to the LaTeX methods, I submit the following justifications...

### ArrayNode LaTeX

Currently, running:

```javascript
math.parse('[x, [y, [z]]]').toTex()
```

returns:

```latex
\begin{bmatrix} x\\ y&\begin{bmatrix} z\\\end{bmatrix}\\\end{bmatrix}
```

which incorrectly typesets to:

![ArrayNode toTex Incorrect](https://user-images.githubusercontent.com/19611724/59390684-ddbdfc00-8d26-11e9-94d3-15b97bd54019.png)

It should instead return:

```latex
\begin{bmatrix} x&\begin{bmatrix} y&\begin{bmatrix} z\end{bmatrix}\end{bmatrix}\end{bmatrix}
```

which correctly typesets to:

![ArrayNode toTex Correct 1](https://user-images.githubusercontent.com/19611724/59390701-e8789100-8d26-11e9-87bc-8de40abe51dc.png)

Interestingly, running the more complicated:

```javascript
math.parse('[x; [y; [z]]]').toTex()
```

returns:

```latex
\begin{bmatrix} x\\\begin{bmatrix} y\\\begin{bmatrix} z\\\end{bmatrix}\\\end{bmatrix}\\\end{bmatrix}
```

which correctly typesets to:

![ArrayNode toTex Correct 2](https://user-images.githubusercontent.com/19611724/59390727-f5958000-8d26-11e9-86f7-dc5416ca1457.png)

However, each final row ends with `\\`, which is unnecessary. A cleaner return value would be:

```latex
\begin{bmatrix} x\\\begin{bmatrix} y\\\begin{bmatrix} z\end{bmatrix}\end{bmatrix}\end{bmatrix}
```

### ObjectNode LaTeX

Currently, running:

```javascript
math.parse('{x: 1, y: 2}').toTex()
```

returns:

```latex
\left\{\begin{array}{ll}\mathbf{x:} & 1\\
\mathbf{y:} & 2\\\end{array}\right\}
```

While this typesets fine in MathJax, it doesn't look quite right in other rendering engines like KaTeX and QuickLaTeX. LaTeX supports a `Bmatrix`, which gives the same brace style, cleaner syntax, and a consistent rendering across engines. Also, the `\n`, spaces between `&`, and final `\\` are unnecessary. The return value would instead be:

```latex
\begin{Bmatrix}\mathbf{x:}&1\\\mathbf{y:}&2\end{Bmatrix}
```

Here are the before/after renderings using MathJax and KaTeX, respectively:

![ObjectNode toTex MathJax](https://user-images.githubusercontent.com/19611724/59390795-2aa1d280-8d27-11e9-8171-81450b8dff60.png)

![ObjectNode toTex KaTeX](https://user-images.githubusercontent.com/19611724/59390803-2fff1d00-8d27-11e9-9341-8baca0a9ecc6.png)

### AccessorNode LaTeX (`dotNotation: true`)

Currently, running:

```javascript
math.parse('O.z').toTex()
```

returns:

```latex
 O.z
```

Since properties within an object are wrapped with `\mathbf{}`, it would maybe look better to wrap accessed properties the same way?:

```latex
 O.\mathbf{z}
```

An example rendering would be:

![AccessorNode toTex KaTeX](https://user-images.githubusercontent.com/19611724/59390845-4907ce00-8d27-11e9-97d7-aea678551bac.png)

### Conclusion

If you agree with the above, the changes I made to the various LaTeX methods should solve it. Please let me know what you think, and sorry for the long comment!

josdejong commented:
Thanks for the detailed explanations Jake. I'm sorry I didn't find time to dive into your PR last weekend. Hope to find time soon.

jakemckown commented:
No worries! Thanks for your time.

josdejong commented:
Thanks for your patience. Really nice job @jakemckown, your implementation looks solid and well tested. I think it will make for a really nice feature (At least I like destructuring a lot in JavaScript).

1. Thanks for your LaTeX fixes, they are very welcome :+1: ! Maybe next time you can create two separate PR's each handing a single topic. 

2. You've put a lot of effort in the unit tests, thanks for that! I see at some places you've extended a test, like `should run forEach on an AssignmentNode without index`. It may be better to split this in two separate tests since the test is becoming very large now. Or is there a special reason to keep it in a single test?

3. About the `// why not new AssignmentNode(node, value)` comment: good point, I think we can simplify this. Can you add this change to your PR?

4. I've been playing around with the array destructuring, it works like expected for matrices of any dimension. I found a few cases that I think we should still look into:
    1. This works:
    ```js
    math.evaluate('[a, b, c] = [1,2,3]')
    ```
    but this doesn't:
    ```js
    math.evaluate('[a, b, c] = [1,2,3] * 2')       // TypeError: ArrayNode expected as "value"
    math.evaluate('[a, b, c] = A', { A: [1,2,3] }) // TypeError: ArrayNode expected as "value"
    ```
    2. I saw that for some reason the command line interface doesn't work yet. Maybe I can do some debugging there (unless you already have an idea what's going on here):

    ```
    ~/mathjs $ ./bin/cli.js 
    > [a,b,c] = [1,2,3]
    TypeError: Cannot read property 'filter' of undefined
    > 
    ```

5. As soon as we offer array destructuring, we can expect the following feature requests (I think we should keep these for a next PR, the current PR is already quite large):
    - destructuring for objects
    - destructuring with `...`, like `[ ...array1, ...array2 ]`

jakemckown commented:
Thanks for the constructive feedback!

1. You're right, the LaTeX changes should have been a separate PR. I'll make sure to keep PR's on-topic in the future!

2. I didn't have a specific reason for extending tests instead of creating new ones, but I like that idea better. I'll extract out the matrix assignment tests in another commit.

3. For sure, I'll commit this change.

4. I'll find some time this week to go through these in depth, but after a first look:

    i. I'm pretty sure the problem is that I'm explicitly requiring that `value` be an `ArrayNode` if `object` is an `ArrayNode`, starting at line 82 of [AssignmentNode.js](https://github.com/jakemckown/mathjs/blob/develop/src/expression/node/AssignmentNode.js):

    ```javascript
    if (isArrayNode(object)) {
      if (this.index) {
        throw new TypeError('"index" not expected when "object" is ArrayNode')
      }
      if (!isArrayNode(this.value)) {
        throw new TypeError('ArrayNode expected as "value"')
      }
      // ...
    ```

    Instead, it should allow any type of `Node` for `value` (like the `OperatorNode` and `SymbolNode` in your examples), which should eventually evaluate to a matrix. I'll fix this.

    ii. I believe this problem stems from the fact that the CLI doesn't allow multiple assignments in a single line, and so it only evaluates the first `Node` in the `blocks` property of any `BlockNode` it encounters (starting at line 214 in [cli.js](https://github.com/jakemckown/mathjs/blob/develop/bin/cli.js)):

    ```javascript
    switch (mode) {
      case 'evaluate':
        // evaluate expression
        try {
          let node = math.parse(expr)
          let res = node.evaluate(scope)

          if (math.isResultSet(res)) {
            // we can have 0 or 1 results in the ResultSet, as the CLI
            // does not allow multiple expressions separated by a return
            res = res.entries[0]
            node = node.blocks
              .filter(function (entry) { return entry.visible })
              .map(function (entry) { return entry.node })[0]
          }
          // ...
    ```

    In the case of your example above, `res` will be a `ResultSet` (with 3 results, not currently allowed!), but `node` will be an `AssignmentNode`, not a `BlockNode`, and so `node.blocks` will be `undefined` (hence the inability to read property `filter`). I'll see what I can do, but is there any particular reason to disallow multiple assignments in the CLI? I don't want to violate a design philosophy here.

5. I really like the idea of destructuring objects and adding a spread operator in mathjs. If you're interested in adding these features, I would be happy to take a shot at it in another PR!

jakemckown commented:
Comment 4.i actually presents an interesting dilemma. Since `value` should be allowed to be any type of `Node`, my implementation should actually be moved to the compile step, not the parse step. So the `assignments` property and `flatten` method are actually part of the problem.

josdejong commented:
I think indeed that the issue 4.i would be solved by evaluating against an actual value instead of a parsed tree.

4.ii that makes sense. The reason that blocks with multiple lines are not supported is because pressing Enter will evaluate the line, it's not possible to have input with multiple lines. It should be fine though to change this into handling multiple lines.

About 5: thanks for your offer! Let's do one step at a time though ;)

jakemckown commented:
Ok, I moved the matrix assignment implementation from the parse step to the compile step. It no longer relies on BlockNode, and evaluating the node returns the result of the last assignment instead of a ResultSet, so it works in the CLI now. You should be able to do things like:

**REPL**
```javascript
math.evaluate('[a, b, c] = [1, 2, 3]')
math.evaluate('[a, b, c] = 2 * [1, 2, 3]')
math.evaluate('[a, b, c] = [1, 2, 3] + [4, 5, 6]')
math.evaluate('[a, b, c] = A', { A: [1, 2, 3] })
math.evaluate('[a, b, c] = A', { A: math.matrix([1, 2, 3]) })
```

**CLI**
```
> [a, b, c] = [1, 2, 3]
3
> A = [4, 5, 6]
A = [4, 5, 6]
> [a, b, c] = A
6
```

What stopped working, however, was accessor assignment like:

```javascript
math.evaluate('B = [16; 17]; [a; B[1,1]; B[2,1]; c] = [18; 19; 20; 21]')
```

 and edge case assignments like:

```javascript
math.evaluate('[a, [b, [c]]] = [10, [11, [12]]]')
math.evaluate('[a; [b; [c]]] = [13; [14; [15]]]')
```

For now, I commented out some unit test assertions for these cases.

Accessor assignment should be easy enough to add, and I think it would be worth the time, but those edge cases may not be worth it. Still, those edge cases do represent valid matrices from a mathematical standpoint, and they are invalid per math.js, because they fail the `validate` function in [utils/array.js](https://github.com/jakemckown/mathjs/blob/develop/src/utils/array.js) when being called from [type/matrix/DenseMatrix.js](https://github.com/jakemckown/mathjs/blob/develop/src/type/matrix/DenseMatrix.js). I'm wondering if maybe it would be beneficial to improve the `validate` function so that it allows all valid matrices such as those above?

jakemckown commented:
Everything works now, including the edge cases above.

You'll see that I removed the `validate` check in [type/matrix/DenseMatrix.js](https://github.com/jakemckown/mathjs/blob/develop/src/type/matrix/DenseMatrix.js) that comes from [utils/array.js](https://github.com/jakemckown/mathjs/blob/develop/src/utils/array.js). I searched every file, and the only file that depends on `validate` is the DenseMatrix. I don't think this check should be done here, because it limits what you can do with a matrix. Matrices without this restriction can be used for things like storing the results of repeated sequential calculations or imported data. It gives you much more flexibility and power.

For functions that depend on matrices having nested matrix elements of equal size, perhaps the `validate` check could be done in those functions?

josdejong commented:
Great, getting there I think, great job :+1: 

I've been playing around with the latest version, works very smoothly! I like the explanatory comments in `AssignmentNode.prototype._compile`, that's really helpful. 

Some feedback points:

1. ArrayNode.toArray: rewrite to `return this.items.map(...)`
2. Can you please leave the line `  '?matrix', // FIXME: should not be needed at all, should be handled by subset` in AssignmentNode as it was? This refactoring should still be done.
3. Instead of `flat = flat || []` you can use a default parameter value like `matchFlatten (keys, values, flat = [])`

4.  The destructuring works fine for DenseMatrices, but doesn't support plain Arrays:
    ```js
    math.evaluate('[a,b,c] = A', { A: [1,2,3] })
    // TypeError: values.toArray is not a function
    at evalArrayAssignmentNode (.../mathjs/lib/expression/node/AssignmentNode.js:182:36)
    ```

5.  Currently, the following is supported:
    ```js
    math.evaluate('[e,f] = [[1,2],[3,4]]')
    ```
    For consistency, should we also support the following too? Or not?
    ```js
    math.evaluate('[e;f] = [[1,2],[3,4]]')
    ```
    Or, should we keep the implementation much simpler by only supporting destructuring of individual cells for example?

6.  The method `AssignmentNode.prototype._compile` has become quite large and complex. I want to have a bit deeper look into this to understand all it does, and think about ways to break it apart or make it simpler. Let's also think critically if there are parts of this new destructuring feature that add a lot of complexity but not that much user value (like my point 5)?

jakemckown commented:
Thanks! The rewrite for `AssignmentNode.prototype._compile` was a bit tricky, but I think it's coming together (see 6 below).

1. Duh, why didn't I use `map` instead of `forEach`? Now that you mention it though, I'm a bit confused about when to use `array.map(cb)` / `array.forEach(cb)` and when to use `map(array, cb)` / `forEach(array, cb)` from [utils/array.js](https://github.com/jakemckown/mathjs/blob/develop/src/utils/array.js). How do `Array.prototype.map.call(array, cb)` and `Array.prototype.forEach.call(array, cb)` make those operations safe?

2. **Fixed**. A previous implementation used the `matrix` function, so I made it mandatory, but it's not anymore. Thanks for catching this.

3. Gotta love ES6.

4. **Fixed**. I added a unit test for this too.

5. The first example, `[e,f] = [[1,2],[3,4]]` is more intuitive, I think. For a destructuring operation, the second example would maybe make more sense as `[e;f] = [[1,2];[3,4]]`, which works. What do you think?

6. The method `AssignmentNode.prototype._compile` is pretty complex, but here's my reasoning for setting it up this way:

### `AssignmentNode.prototype._compile`

The basic structure of this method is:

```javascript
return delegate(this.object, this.index, this.value)

function delegate (object, index, value, scope, args, context) {...}

// utility functions
function getEvalArrayAssignmentNode (object, index, value) {...}

function getEvalSymbolAssignmentNode (object, index, value, scope, args, context) {...}

function getEvalPropertyAssignmentNode (object, index, value, scope, args, context) {...}

function getEvalSubsetAssignmentNode (object, index, value, scope, args, context) {...}

function getEvalPropertySubsetAssignmentNode (object, index, value, scope, args, context) {...}
```

`delegate` is first called with only `object`, `[index]`, and `value`. All three arguments, if provided, should be some type of `Node`. The usual operation happens: The `Node` types are checked to determine which return statement is issued, but now the original implementations of each return statement are extracted out into individual utility functions.

The basic structure of each utility function, with the exception of `getEvalArrayAssignmentNode` is:

```javascript
if (scope) {
  // 'value' is an actual value, not a Node; return the evaluation...
} else {
  // 'value' is a Node; return a function of the form: function evaluate (scope, args, context) {...}...
}
```

If `object` is anything but an `ArrayNode`, everything happens as it did before. If `object` is an `ArrayNode`, then `getEvalArrayAssignmentNode` is returned, within which `delegate` is called again. But this time there is a `scope`, and `value` has been evaluated to a `DenseMatrix` or `Array` using `scope`, and further coerced into an `Array` of values to be matched against each `Node` in `object` using `matchFlatten`. The result of `matchFlatten` looks something like:

```javascript
matches = [
  { key: SymbolNode, value: 1 }
  { key: AccessorNode, value: 2 }
  // ...
]
```

For each pair of `key` and `value` in `matches`, `delegate` gets called a second time, but this time `scope` gets passed along as well. `object` and `[index]` should still be a `Node`, but since `value` is an actual value now, we use `scope` to further compile down `object` and `[index]` and make our assignments.

Lastly, the final result of these assignments is returned from the very first call to `delegate`, and `scope` should now contain all assignments made.

I am completely open to any challenges to this implementation, so that it can be made as simple and reasonable as possible, but I will need some help to do that, because there are already a lot of functions returning functions returning functions =). The two biggest barriers are:

1. As we discussed before, `value` can be any type of `Node` (which eventually evaluates to a `matrix` or `array`), so to get a 1:1 assignment to `object`, `value` must first be evaluated.

2. `object` cannot be evaluated without a `scope`, which is why there are two passes through `delegate`, the first to evaluate `value`, the second to make the assignments.

josdejong commented:
Thanks for the improvements :+1: 

**About the actual behavior**

The matrix `[[1,2];[3,4]]` results in a 3d matrix with size 2x1x2, `[[1,2],[3,4]]` is a 2x2 matrix. It's probably more intuitive to talk about `[1,2;3,4]` :).

What I meant is the following: currently, your implementation works with nested arrays. From a mathematical point of view though, it may make more sense to work with rows and columns. 

Let's take the following example to make sure we're on the same page. Create a matrix with 2 rows and 3 columns:

```
A = [1,2,3;4,5,6]  # this is the same as [[1,2,3],[4,5,6]]
```
![image](https://user-images.githubusercontent.com/568626/60959533-a41de800-a308-11e9-8217-c82665618328.png)

Now, I would expect to be able to extract the 2 rows like this:

![image](https://user-images.githubusercontent.com/568626/60959620-cf083c00-a308-11e9-9fe2-8d0468e3bc92.png)

```
[r1;r2] = [1,2,3;4,5,6]
# I would expect here to get the rows:
# r1 = [1,2,3]
# r2 = [4,5,6]
```

And to extract the 3 columns:

![image](https://user-images.githubusercontent.com/568626/60959706-fa8b2680-a308-11e9-9f1f-2a0ae670b9ae.png)

```
[c1, c2, c3] = [1,2,3;4,5,6]
# I would expect here to get:
# c1 = [1;4]
# c2 = [2;5]
# c3 = [3;6]
```

This is a totally different approach though. What do you think?



**About the _compile code**

Thanks for the explanation. The code of looks `AssignmentNode.prototype._compile` looks well structured, I can follow what it does, but it seems that it currently does two things mixed. But maybe I don't fully understand it.

What helps me personally to better understand the code is to rename `getEval*` to `compile*` (i.e. `getEvalSymbolAssignmentNode` becomes `compileSymbolAssignmentNode`), and rename `delegate` to `compileAny`. Is that an idea for more clear naming?

I see the most of the `getEval*` functions handle two cases: when there is no `scope`, they return a function that can be invoked to evaluate the expression, but when there is a `scope`, they return a result. It feels to me that these are two completely different things (compile and evaluate) that should not go in a single function. I think it will help to split this in separate functions, and have `AssignmentNode.prototype._compile` call the "compile" part of it, and have `evalArrayAssignmentNode` call the "evaluate" part of the code including `scope`.

jakemckown commented:
Thanks!

**About matrix assignment behavior**

Firstly, I really like your idea of being able to do things like:

![r1, r2](https://user-images.githubusercontent.com/19611724/60997439-ac2b5700-a30b-11e9-84d2-c0d924988e4c.png)
![r1](https://user-images.githubusercontent.com/19611724/60997453-b3526500-a30b-11e9-8f78-33782dec8126.png) ![r2](https://user-images.githubusercontent.com/19611724/60997486-c06f5400-a30b-11e9-8c2e-6abedc03c487.png)

and:

![c1, c2, c3](https://user-images.githubusercontent.com/19611724/60997539-e1d04000-a30b-11e9-9937-2f0397143668.png)
![c1](https://user-images.githubusercontent.com/19611724/60997566-f4e31000-a30b-11e9-9e98-ea6b611e3dfe.png) ![c2](https://user-images.githubusercontent.com/19611724/60997579-fb718780-a30b-11e9-8d62-b864222d2b91.png) ![c3](https://user-images.githubusercontent.com/19611724/60997586-01676880-a30c-11e9-9dcc-ec12329abb1f.png)

You're right that it's not currently supported by my implementation, but I think it probably should be. It's essentially a shortcut for doing range operations followed by assignments!

It would require an additional step to, say if the number of rows **and** columns are not equal on both sides, check whether the number of rows **or** number of columns is equal. But it could be useful as an elegant way to slice a matrix (like the result of a matrix-returning function) into smaller matrices for separate processing?

Regarding the parser syntax for describing a matrix, I think this is where we disagree. You said:

```
A = [1,2,3;4,5,6]  # this is the same as [[1,2,3],[4,5,6]]
```

While I agree that `[1,2,3;4,5,6]` should describe:

![2x3](https://user-images.githubusercontent.com/19611724/60998502-eb5aa780-a30d-11e9-8dbd-652168ebc13b.png)

I would argue that `[[1,2,3],[4,5,6]]` should describe:

![1x2 with nested 1x3s](https://user-images.githubusercontent.com/19611724/60998594-1513ce80-a30e-11e9-91b4-8868f440f0f0.png)

I know they are both currently interpreted by the parser as the same, and I understand that the first matrix object's `_data` property is an array of `[[1, 2, 3], [4, 5, 6]]`, but I would warn against mixing JavaScript array syntax and math.js matrix syntax in the parser, as it can make some expressions ambiguous, and others impossible. For example, there is currently no way (that I know of) to obtain the second matrix from the parser! And yet it is a perfectly valid matrix, capable of recursive operations like:

![1x2 with nested 1x3s operation](https://user-images.githubusercontent.com/19611724/60999150-4d67dc80-a30f-11e9-87c4-b53e94ee628f.png)

If the parser syntax were modified so that `,` always designates column separation, `;` always designates row separation, and `[` and `]` always designate the boundaries of a matrix, then I think matrices of any form should be describable? Additionally, the `_data` property of both matrices could remain as an array of `[[1, 2, 3], [4, 5, 6]]`, but the `_size` property could maybe be used to distinguish between the two (`[2, 3]` and `[1, 2]`, respectively). What do you think?

**About the _compile code**

Renaming `getEval*` to `compile*` and `delegate` to `compileAny` is probably better. I can also separate out the two halves of each function into separate functions as well. I only put them together because each half shares some of the initial code in the function, and I didn't want to repeat code. But what you're talking about is probably easier to reason with.

For your last sentence, do you mean to pull `evalArrayAssignmentNode` completely out of `_compile` as a standalone function? Let me know what you prefer, and I'll make the change =).

josdejong commented:
First, sorry again for the late response Jake, I sometimes struggle to keep up with all PR's, issues, and general maintenance. This PR requires some thinking through.

> Regarding the parser syntax for describing a matrix, I think this is where we disagree.

Indeed. This is the current implementation of how things work in mathjs. I'm always open to rethinking decisions though :). I think we do have two use cases here:

1. Create a multi dimensional matrix. This can currently be done in two ways:
    - Create a 2d matrix like `[1,2,3;4,5,6]`
    - Create an n-d matrix like `[[[1,2],[3,4]],[[5,6],[7,8]]]`
2. Create a 1d matrix containing 1d matrices, ie. your expectation of what `[[1,2,3],[4,5,6]]` should result in. This cannot really be done in the expression parser (at least not using the `Matrix` classes. You can achieve it when working with plain `Array` but you have to jump through hoops).

Mathjs is a a bit inconsistent in that it does not create 2d matrices by default for inputs like `[1,2,3]`, something that for example Matlab does very consistently.

The reason that mathjs doesn't support use case (2) is that this isn't a very common need, and if you need it, it may be a code smell. Normally you do want use case (1): make it easy to create a matrix with mutliple dimensions. This is also why one of your previous examples, `[x; [y; [z]]]` feels quite artifical to me, and why I explicitely want to think about whether we do want to support that: I can't come up with a real world use case for it. Do you have a concrete use case for wanting to create a 1d matrix containing 1d matrices, or have to deal with something like `[x; [y; [z]]]`?

Maybe it helps if we try to list use cases for destructuring, so we get a more clear view on what we actually want to support? The use cases that I would liek to use it for are:

1. slice/split a matrix in individual rows or columns
2. destructure pairs like `[longitude, latitude] = location` or `[x, y, z] = point` 


> For your last sentence, do you mean to pull `evalArrayAssignmentNode` completely out of _compile as a standalone function? Let me know what you prefer, and I'll make the change =).

For me it's not necessary to move `evalArrayAssignmentNode` outside of `_compile`, as long as it only does a single thing (i.e. evaluate something). So far basically all `_compile` methods return an "evaluate" function, like the old structure was:
```js
AssignmentNode.prototype._compile = function (math, argNames) {
  ...
  return function evalAssignmentNode (scope, args, context) {
    ...
  }
  ...
}
```

jakemckown commented:
No apologies needed, Jos. Thanks for getting back to me when you can.

**_compile Method**

I renamed all the `getEval*` functions to `compile*`, and separated out the `compile*` and `eval*` functions, keeping all the `eval*` functions that come from `compileArrayAssignmentNode` entirely within that parent function. Let me know what you think.

**Destructuring**

I'll admit that `[x; [y; [z]]]` looks a bit farfetched from a symbolic/numerical perspective, but from a destructuring perspective it's just an artificial example that demonstrates a pattern I use often in Mathcad with 1-d matrices containing 1-d matrices. Here's a use case that extends your `[x, y, z] = point` example above:

Assume some function that returns a list of points, represented as `P`:

![P](https://user-images.githubusercontent.com/19611724/61597964-e6c6a680-abcb-11e9-8ef5-dee86cb37a28.png)

implemented as:

```
P = [[1, 2, 3]; [4, 5, 6]; [7, 8, 9]]
```

***
**Quick note:** Although this looks like a 1-d matrix containing 1-d matrices, in Mathcad this is actually a 2-d matrix (with 1 column) containing 2-d matrices (with 1 row), because all matrices in Mathcad are represented with 2-d arrays as follows:

```c
// complex array type
typedef struct tagCOMPLEXARRAY  {
    unsigned int rows;
    unsigned int cols;
    double **hReal; // hReal[cols][rows],  == NULL when the real part is zero 
    double **hImag; // hImag[cols][rows],  == NULL when the imaginary part is zero
} COMPLEXARRAY;
```

This yields the following differences in internal structures:

| Expression | math.js | Mathcad |
| --- | :-: | :-: |
| `[1, 2, 3]` | `[1, 2, 3]` | `[[1], [2], [3]]` |
| `[1; 2; 3]` | `[[1], [2], [3]]` | `[[1, 2, 3]]` |

I'm not sure what the performance implications are here; I just thought I would mention it in case this information can be useful for you.
***

Anyway, instead of destructuring `P` in four steps like:

![Destructuring P into points](https://user-images.githubusercontent.com/19611724/61598031-9dc32200-abcc-11e9-9f52-c169b8fcc470.png) ![Destructuring p1](https://user-images.githubusercontent.com/19611724/61598032-a3206c80-abcc-11e9-8f45-3916413ca6d2.png) ![Destructuring p2](https://user-images.githubusercontent.com/19611724/61598034-a7e52080-abcc-11e9-8f42-f3efb768967a.png) ![Destructuring p3](https://user-images.githubusercontent.com/19611724/61598036-addb0180-abcc-11e9-9153-3f24f0f5a5b5.png)

implemented as:

```
[p1; p2; p3] = points
[x1, y1, z1] = p1
[x2, y2, z2] = p2
[x3, y3, z3] = p3
```

With 1-d matrices in a 1-d matrix, we can destructure it in one step like:

![Destructuring P into coordinates](https://user-images.githubusercontent.com/19611724/61598049-d06d1a80-abcc-11e9-8c68-add28f5ff699.png)

implemented as:

```
[[x1, y1, z1]; [x2, y2, z2]; [x3, y3, z3]] = P
```

Now imagine a possible need to break one point into coordinates while keeping the other points whole, like `[p1; p2; [x3, y3, z3]]`, and we're starting to get something that looks a lot more like `[x; [y; [z]]]`.

The above use case can be demonstrated in the current implementation with:

```
// bin/cli.js

> P = [[1, 2, 3]; [4, 5, 6]; [7, 8, 9]]
P = [[[1, 2, 3]], [[4, 5, 6]], [[7, 8, 9]]]
> [[x1, y1, z1]; [x2, y2, z2]; [x3, y3, z3]] = P
9
> x1
1
> z3
9
```

**Slicing Matrices**

Since this PR is already pretty large (that's my fault!), and since it achieves what I had originally intended, what are your thoughts on the following path forward?:

1. Keep the current implementation as-is (with any additional comments from you about the structure of `AssignmentNode.prototype._compile`.
2. Create a new `matrix-slicing` branch (or whatever name you like) for expanding destructuring to include row and column slicing.

josdejong commented:
**Code**

The splitting of compile and evaluate works out beautifully! Thanks for the update Jake!

Since `compileAny` is now only evaluated once inside `AssignmentNode.prototype._compile`, it's now a needless indirection and you could also directly execute it's contents in `AssignmentNode.prototype._compile` :). It's fine with me though if you like it more this way.

**Matrices containing matrices?**

Before we can merge this PR though we have to be aligned on its behaviour, since once introduced, it's hard to change course. Thanks for explaining how you use this in MathCad, that really helps me understand where you come from. You're really thinking about matrices containing matrices, whilst I have in mind destructuring elements or chunks of a single matrix.

We have to align the behaviour with how mathjs works and what users will expect. In mathjs, nested arrays are interpreted as a single matrix with multiple dimensions. Creating a nested array in the expression parser like `[[1, 2], [3,4]]` is a notation to create a dense matrix with multiple dimensions (2x2 in this example). It will create a single matrix, not a 1d matrix containing 1d matrices. It's sort of impossible to create a matrix containing a matrix: if you do something like `A = [1,2,3]; A[4] = [5,6,7]` you get an error `TypeError: Scalar expected`. 

So in short: the model of mathjs is having multi-dimensional matrices. It has no support for matrices containing matrices.

**Behaviour**

I think the current implementation of destructuring largely does what we want. For destructuring individual cells it just works. But we have to decide about how to deal with rows and columns. 

Since mathjs doesn't support matrices containing matrices, I would expect symmetric behaviour in https://github.com/josdejong/mathjs/pull/1542#issuecomment-509994217: be able to destructure either rows or columns in exactly the same way. The destructuring right now relies on the internal structure of a DenseMatrix (which happens to use nested arrays internall), but should be independent from that I think (like SparseMatrix has a different internal structure).

So I would expect (summarizing [my earlier comment](https://github.com/josdejong/mathjs/pull/1542#issuecomment-509994217)):

```js
[r1;r2] = [1,2,3;4,5,6]
# r1 = [1,2,3]
# r2 = [4,5,6]

[c1, c2, c3] = [1,2,3;4,5,6]
# c1 = [1;4]
# c2 = [2;5]
# c3 = [3;6]

# This looks really odd though when using the nested array notation, that's a down side:
[r1;r2] = [[1,2,3],[4,5,6]] 
# whut? why not [r1,r2] = [[1,2,3],[4,5,6]] 
```

I understand you could do 

**Options**

I think we have a couple of options:

1. Keep the current implementation which does `[r1,r2] = [[1,2,3],[4,5,6]]`. 
2. Implement my proposal with symmetric behavior for rows/columns.
3. Only implement destructuring of individual cells for now, not of rows/columns. Like your example `[[x1, y1, z1]; [x2, y2, z2]; [x3, y3, z3]] = P`

I'm strongly in doubt about (1) because it is at odds with the current behavior and relies on the internal structure of a DenseMatrix. I'm in doubt about (2) too, because it doesn't look logical in all cases. If we go for (3) there are no difficulties at all, we can go for that.

Whether we want to be able to destructure rows/columns depends on the use cases we foresee. For my personal needs, a `slice` function instead would be just as helpful and very straightforward.


**Use cases**

I'm still looking for actual use cases from your side :)

You show an example `[p1; p2; [x3, y3, z3]]`, but this feels really artificial to me, I can't imagine a real use case where I would want to write code this way. As long as I don't see convincing use cases I'm in double whether we should implement support for it.

Can you give some use cases of where/how you utilize working with matrices containing matrices? (I normally use objects to hold multiple matrices, never do nesting) Do you have concrete examples how you use destructuring?

jakemckown commented:
Thanks, Jos.

All of the concrete examples of destructuring I could give you would be from Mathcad worksheets (the behavior of which I'm trying to mimic with this PR), which I've developed for the company I work for. Unfortunately, I don't have ownership of those worksheets and cannot provide them here. But the point example above is very similar and not at all artificial. Many of the use cases involve destructuring a matrix being returned from a function, so an object wouldn't work for that.

Regarding the current implementation's dependence on DenseMatrix, I'm not sure how to remove that. The right-hand side of an assignment can be any arbitrary expression. You don't know what the structure of the resulting matrix or array looks like until you evaluate the expression.

harrysarson commented:
> Only implement destructuring of individual cells for now, not of rows/columns. Like your example [[x1, y1, z1]; [x2, y2, z2]; [x3, y3, z3]] = P

If I may weigh in, I love the destructuring idea. I would vote for Jos's third option. It would be possible, going forward, to extend the implementation to either option 1, 2 or nested matrices. 

Option 3 is a solid and conservative start though si it gets my vote :slightly_smiling:

josdejong commented:
@jakemckown too bad you can't share your examples. We discovered already that the core behavior of MathCad and mathjs differs, that's probably why copying this MathCad feature into mathjs is hard. I don't know whether MathCad can work with objects. Since this is so natural in JavaScript and mathjs to work with objects, this may lead to very different coding patterns making the need to work with matrices inside matrices redundant in the first place. Without concrete examples it's hard to tell though.

Jake, what is your opinion about the three options I listed? Like @harrysarson I think we can best go for the conservative approach (option 3) at this point. 

> Regarding the current implementation's dependence on DenseMatrix, I'm not sure how to remove that. 

My feeling is that it may be not trivial to make the implementation independent of the DenseMatrix implementation. If we go for the conservative option 3 though, it may be very easy (check whether `.size()` is the same, then use `.forEach()` to loop over each cell and apply an assignment). I guess we should first agree on which solution we go for though :)

josdejong commented:
@jackcmason is this PR still on your radar? I would love to hear your opinion about just going for option 3 for now. I'm not sure if that would address your needs or make the solution unusable for your use case.

jakemckown commented:
Jos, I'm assuming by option 3 you are referring to an `AssignmentNode` where both the `object` and `value` properties are an `ArrayNode`? This was my very first implementation, but it does not take into account that the right-hand side of an assignment can be any expression, meaning that `value` may not be an `ArrayNode`. Without first evaluating the right-hand side of the assignment to a `DenseMatrix` or `Array`, there is no way to match the cells of each side.

My current implementation already fits option 3, but depends on `DenseMatrix`. So the real options are:

1. Keep the current implementation as-is.
2. Go back to the `ArrayNode`-only implementation.

I'm okay with option 1, but option 2 doesn't fit my use case.

josdejong commented:
> Jos, I'm assuming by option 3 you are referring to an AssignmentNode where both the object and value properties are an ArrayNode?

No, with option 3 I mean the current implementation, but adding some restrictions, namely: we only allow left hand side and right hand side when they have equal dimensions. This means you can only destructure individual cells (for example `[a,b;c,d]=[1,2;3,4]`), but not entire rows or columns  (for example `[a,b]=[1,2;3,4]`) since it's ambiguous how to interpret it (will the example destructure columns or rows?).

jakemckown commented:
The current implementation’s `matchFlatten` function checks that both sides have equal dimensions and throws an error otherwise. It doesn’t allow ambiguity like `[a,b]=[1,2;3,4]`.

josdejong commented:
> It doesn’t allow ambiguity like `[a,b]=[1,2;3,4]`.

I'm not sure if I understand you, the current implementation does allow evaluating `[a,b]=[1,2;3,4]`. In this example the size of the left hand side matrix is `[2]` and the size of right hand side matrix is `[2,2]`, so these matrices do not have equal dimensions.

The code is of course evaluates deterministic :). With "ambiguity" I mean that a user can have different expectations about what this expression will evaluate to. Similar to implicit multiplication that creates ambiguity: is `1 / 2 x` evaluated as `(1 / 2) * x` or as `1 / (2 * x)`?

I tried to explain the differences we have in expected behavior in this comment: https://github.com/josdejong/mathjs/pull/1542#issuecomment-518987979, with as conclusion that it may be wise to be conservative for now (option 3) and only allow destucturing of matrices with equal dimensions/size. What you you think about it?

gwhitney commented:
If this PR is ever revived (or if all of these comments are turned into a Discussion detailing what the behavior should be if it is reimplemented), then I have the following suggestion to guide the semantics of destructuring assignment in general:

* Proposal: if 'E(x,y,z)' is an expression with variables 'x', 'y', and 'z', then the assignment `E(x,y,z) = W` only results in assigning the values X, Y, and Z to the variables 'x', 'y', and 'z' if the expression 'E(X,Y,Z)' obtained by substituting the values X, Y, and Z respectively for the variables 'x', 'y', and 'z' into the expression 'E(x,y,z)' would evaluate to W. (And of course the same thing for different numbers of variables.)

Applying this proposal to destructuring blocks of matrices, `[r1, r2] = [1,2,3;4,5,6]` should result in `r1 = [1,2,3]` and `r2 = [4,5,6]` because `[[1,2,3], [4,5,6]]` evaluates to the identical matrix as `[1,2,3;4,5,6]`. On the other hand, `[r1;r2] = [1,2,3;4,5,6]` should **not** result in these same assignments because, as you point out, `[[1,2,3];[4,5,6]]` does not evaluate to this matrix, but rather the 2x1x3 matrix `[[[1,2,3]], [[4,5,6]]]`. In fact, I do not believe there are any values for r1 and r2 such that `[r1;r2]` evaluates to `[1,2,3;4,5,6]`, so this should just fail. But `[r1;r2] = [1;4]` _should_ work to set `r1 = 1` and `r2 = 4` because of course `[1;4]` is equal to `[1;4]`.

Another corollary is that unless there is a notation that assembles a matrix from its columns, there is no way to destructure a matrix into its columns. Correct me if I am wrong, but at the moment the only way to assemble a matrix from columns is to write `concat(c1,c2,c3)`. So accepting the proposal would mean that there would only be a way assemble a matrix from columns if mathjs is willing to support an expression like `concat(c1,c2,c3) = [1,2,3;4,5,6]` resulting in `c1=[1;4]`, `c2=[2;5]`, and `c3=[3,6]`. Or conceivably mathjs could introduce a notation like `[1;4|2;5|3;6]` that _does_ assemble a matrix from columns so that  then `[c1|c2|c3] = [1,2,3;4,5,6]` would result in those same assignments.

Anyhow, I think that this proposal results in zero surprises in the results of destructuring assignments in general, not limited just to matrices/arrays, which is why I endorse it.

josdejong commented:
Ah, yes, this is a very nice feature. To bad it got stalled.

@gwhitney I think your simple and straightforward proposal makes sense. It depends though how we look at it. If we look at it from a JavaScript perspective and see a matrix as a nested array, we could extract the nested parts from the outer array indeed. However, I think it may make more sense to look at this from a mathematical perspective, with the concept of a 2d matrix having rows and columns. 

I think you have two cases give a matrix `[1,2;3,4]` (which equals `[[1,2],[3,4]]`):

- Split rows (`;`): 
  `[r1; r2] = [1,2;3,4]` should result in `r1=[[1,2]]` and `r2=[[3,4]]`. 
  Assemble: `concat(r1, r2, 1)` gives the original matrix again
- Split columns (`,`): 
  `[r1, r2] = [1,2;3,4]` should result in `r1=[1;3]=[[1],[3]]` and `r2=[2;4]=[[2],[4]]`. 
  Assemble: `concat(r1, r2, 2)` gives the original matrix again

Does that make sense?

gwhitney commented:
I was looking at this from _neither_ a mathematical or JavaScript perspective, but rather from the point of view of the notation/language that mathjs itself parses and supports. I personally would find it very confusing if `[c1,c2,c3] = [1,2,3;4,5,6]` assigned the three columns to the variables c1, c2, and c3, **but** with those values the expression `[c1, c2, c3]` did not then reconstruct that same matrix `[1,2,3;4,5,6]`. **Every** example of destructuring assignment that I know of in any language that supports such things satisfies this sort of "restructuring invariance" property. So I would _strongly_ recommend against adopting any destructuring assignment syntax that doesn't have this property -- I would much prefer not having it at all, or in the particular case of matrix destructuring, limiting it to only destructurings down to the individual entry level (as I think is proposed above), which _would_ have the invariance property.

All that said, I personally don't see any problem with supporting any destructuring assignment which satisfies the invariance; at the moment, that would mean `[r1, r2] = [1,2,3;4,5,6]` could work, since if `r1 = [1,2,3]` and `r2 = [4,5,6]`, then the mathjs expression `[r1, r2]` does indeed return the matrix `[1,2,3;4,5,6]`. Note that this actually has nothing to do with the (coincidental) internal representation of Arrays/Matrices as JavaScript entities; it's purely about the syntax and semantics of the mathjs expressions. And it would mean that exactly what notations mathjs adopts in its parsed language for _building_ matrices would determine which sorts of destructurings one could do, which in my mind is exactly what "destructuring" does and should mean.

Thanks for your consideration of these points, which (obviously) I see as very important.

josdejong commented:
Thanks, I understand your point. 

Maybe I try to make destructuring too powerful, wanting to be able to split a matrix apart in either rows or columns. 

It makes sense to look from a syntax point of view and make sure that there is "restructuring invariance". If we go with that approach, it boils down to be able to split rows of a 2d matrix, or individual values out of a 1d matrix. I like the simplicity of it. I'm not sure if it covers the typical use cases when only supporting splitting of rows, I feel additional utility functions coming up to cover other cases, but indeed maybe we should create separate functions for that instead of overloading the destructuring operation.

Since we will not work this out right now, I suggest we postpone a decision until we actually go implement it. (in time, we may collect a tiny little bit more wisdom and may be able to make a better decision 😉).

josdejong commented:
For now, I'll close this PR. Let's reopen or create a new one when we start actively working on it again.
